cep-cache-output.md

```diff
@@ -0,0 +1,103 @@
+
+There are many downsides to the behavior of `conda-build`: it's very implicit, hard to understand and hard to debug (for example, if an output is defined with the same name as the top-level recipe, this output will get the same requirements attached as the top-level).
+
+For the v1 spec we are attempting to formalize the workings of the "top-level" build. For this, we introduce a new top-level key `cache`, that has the same values as a regular output.
+
+## Specification
+
+The top-level `cache` key looks as follows:
+
+```yaml
+cache:
+  requirements:
+    build:
+      - ${{ compiler('c') }}
+      - cmake
+      - ninja
+    host:
+      - libzlib
+    # run: ...
+    # run_constraints: ...
+    # ignore_run_exports: ...
+  build:
+    # note: in the current implementation, all other keys for build
+    #       are also valid, but unused
+    script: build.sh
+```
+
+<details> 
+  <summary>script build.sh default discussion</summary>
+We had some debate wether the cache output should _also_ default to `build.sh` or should not have any default value for the `script`. This is still undecided.
+</details>
+
+When computing variants and used variables, rattler-build looks at the union of a given output and the cache. That means, even if an output does not define any requirements, the cache would still add a variant for the `c_compiler`.
+
+When rattler-build executes the recipe, it will start by building the cache output that is appropriate for the current variant. This is computed by looking at all "used-variables" for the cache output and computing a "hash" for the cache. The build itself is executed in the same way as any other build.
+
+The variant keys that are injected at build time is the subset used by the cache output.
+
+When the cache build is done, the newly created files are moved outside of the `host-prefix`. Post-processing is not performed on the files beyond memoizing what files contain the `$PREFIX` (which is later replaced in binaries and text files with the actual build-prefix with a simple 1-to-1 replacement as the prefix-length are guaranteed to be equal).
+
+The cache _only_ restores files that were added to the prefix (conda-build also restored source files). Sources are re-created from scratch and no intermediate state is kept.
+
+<details> 
+  <summary>**Workaround for keeping intermediate build state**</summary>
+In order to keep state from the build around in the outputs, one could exercise the build in the $PREFIX and ignore the files in the outputs, or the build folder coule be placed in the `/tmp` folder:
+
+```bash
+cmake -B /tmp/mybuild -S .
+cmake --build /tmp/mybuild
+```
+
+Since this is not extremely convenient, we could discuss adding a "scratchpad" outside the `src` dir to rattler-build that will be kept between outputs and removed at the end of the whole build.
+</details>
```

Comment:
To double-check, this would save `$SRC_DIR` to the scratchpad in between operations, and restore it for the installation of any output that needs it?

I would go as far as saying that `cache:` implies this behaviour. I don't want to cache stuff in `$PREFIX` (then I could just install it and call that an output), I want to cache it _without_ installation.